node/node-netbox.inc.php

```diff
@@ -0,0 +1,182 @@
+		curl_setopt($url, CURLOPT_CONNECTTIMEOUT,5);
+		curl_setopt($url, CURLOPT_RETURNTRANSFER,true);
+		curl_setopt($url, CURLOPT_CAPATH,"/etc/grid-security/certificates");
+		$out = curl_exec($url);
+		if (curl_error($url)) {
+			echo curl_error($url);
+			echo "curl error";
+			return null;
+		}
+		curl_close($url);
+		
+		$json_out = json_decode($out,true);
+		
+		// Do not have a count value or a netbox id - something is wrong
+		if (!isset($json_out['count']) && !(isset($json_out['id']))) { 
+			return null;
+		}
+		
+		// We have an id so just return the json object as it was a direct request
+		if (isset($json_out['id'])) {
+			return $json_out;
+		}
+		
+		if (isset($json_out['count'])) {
+			if ($json_out['count'] == 0) {
+				// Search returned no values so return null
+				return null;
+			} else {
+				// Search returned multiple values so return the array
+				return $json_out['results'];
+			}
+		}
+	}
+	
+	// Get Netbox info for a single machine
+	// To the Netbox API this is technically a search
+	// So be careful we only get one response back	
+    private function get_netbox_info($machineName)
+    {
+        global $SQL;
+
+		$query=$this->netbox_search("/dcim/devices/",array("name"=>$SQL->real_escape_string($machineName)));
+		if ($query == null) {
+			return null;
+		} else if (count($query) > 1) {
+			echo "Got multiple results from Netbox!";
+			return null;
+		} else {
+			return $query[0];
+		}	
+    }
+    
+	// Get Netbox info using netbox id
+	private function get_netbox_info_by_id($id) {
+		// Trailing / required to avoid redirecting
+		return $this->netbox_query("/dcim/devices/", $id . "/");
+	}
+		
+	// Run a generic Netbox query against an API path,
+	// Takes an array of search terms	
+	private function netbox_search($path,$params) {
+		return $this->netbox_query($path, '?' .  http_build_query($params));
+	}
+			
+    private function render_pdu_list($rack_pdus)
+    {
+        echo "<h3>Rack Power</h3>\n";
+		
+        if ($rack_pdus != null) {
+            echo "<dl>\n";
+            foreach ($rack_pdus as $rack_pdu) {
+				$conns=$this->netbox_search("/dcim/power-connections/", array("device"=>$rack_pdu['name']));
+				if ($conns != null) {
+					// For situation where downstream power ports are configured, but upstream is not
+					$found_supply=false;
+					foreach ($conns as $conn) {
+						// Device is the device receiving power, power_outlet device is the device supplying power
+						if ($conn['device']['id'] == $rack_pdu['id']) {
+							$found_supply=true;
+							$room_pdu=$this->get_netbox_info_by_id($conn['power_outlet']['device']['id']);
+		                	printf("<dt>%s (%s)</dt><dd>%s</dd>\n", $room_pdu['name'], $rack_pdu['name'], $room_pdu['site']['name']);
+						}
+					}
+					if ($found_supply == false) {
+						printf("<dt>Magic! (%s)</dt><dd>Unknown</dd>\n", $rack_pdu['name']);	
+					}
+				} else {
+					printf("<dt>Magic! (%s)</dt><dd>Unknown</dd>\n", $rack_pdu['name']);
```

Comment:
What does "Magic!" mean in these two contexts?
Is it worth distinguishing between them?

Reply:
I've clarified their meanings